### DISCUSSION TITLE
Assigning multiple material properties to diffused interfaces

### DISCUSSION THREAD

**Discussion by MinMorsh** [idaholab/moose#29610] [Q&A Modules: Phase field]
I am using Voronoi tessellation (PolycrytalVoronoi) to create a 64 grain model (op_num = 20) in 3D and then assigning the electrical conductivities (bulk and interface) using bnds variable (BndsCalcAux). Here, when bnds>0.9, I am assigning bulk property and interfacial property to the rest. 
Here is the Materials block of the current model that I have. 

```
[Materials]
  [./ecCond] #The equation defining the electrical conductivity is defined here, using two ifs
      type = ParsedMaterial
      block = 0
      constant_names = 'length_scale k_b k_int'  # b = bulk, int = interface
      constant_expressions = '1 4.9758e-7 6.63e-5'
      function = 'sk_b:= length_scale*k_b; sk_int:= length_scale*k_int; if(bnds>0.90,sk_b,sk_int)'
      outputs = exodus
      property_name = ecC
      coupled_variables= 'bnds'
    [../]
[]
```

I need to assign multiple materials to the interface so that some part of the interface should have one type of interfacial material (less conductive, denoted as sk_int1), and other part of the interface should have another type (more conductive, denoted as sk_int2). Any idea how to achieve that?

**Comment by GiudGiud** [comment]
> so that some part of the interface has one type of interfacial material (less conductive, denoted as sk_int1), and other part of the interface has another type (more conductive, denoted as sk_int2). 

how is the interface defined? Is it a sideset in the mesh? Is it a subdomain?
Is there a criterion already for selecting between the types?

**Reply by MinMorsh** [reply]
The diffused interface is defined using the int_width parameter in the PolycrystalVoronoi user object. The bnds variable takes a value of 1 at the grains and less than 1 at the diffused interfaces. This setup is used to assign two properties—grains (e.g., Si) and interfaces (e.g., Al₂O₃). In my case, I need to assign one material (e.g., Si) to the grains and two different materials (e.g., Al₂O₃ and ZrO₂) to the interfaces, but at specific locations within the interfaces.

![3D_64gr](https://github.com/user-attachments/assets/4dc0f5be-ab7a-484b-af46-61c74bae61c4)

**Reply by GiudGiud** [reply]
> but at specific locations within the interfaces.

how do you determine those? Using the bnds field? Using a separate subdomain?

**Reply by MinMorsh** [reply]
That's what I am looking for here. Can I make separate subdomains of the interfaces so that I can assign different properties to them?

**Reply by GiudGiud** [reply]
if the mesh is fine enough to resolve them, sure. You can use this object to modify the subdomains based on a variable value:

https://mooseframework.inl.gov/source/meshmodifiers/CoupledVarThresholdElementSubdomainModifier.html

**Reply by MinMorsh** [reply]
I have been using similar technique here `function = 'sk_b:= length_scale*k_b; sk_int:= length_scale*k_int; if(bnds>0.90,sk_b,sk_int)'` (using if statement) to assign properties to grains and interfaces. Now I am looking for something that will enable me to assign multiple properties to the interfaces/GBs (illustrated below). 
![Picture1](https://github.com/user-attachments/assets/d445422f-fcc6-443c-acb7-110ca2ff8b09)

**Reply by MinMorsh** [reply]
Yes, the mesh is quite refined at the interfaces. Below is the current mesh I am using, generated with Adaptive Mesh Refinement (AMR).

![image](https://github.com/user-attachments/assets/b85c40b4-7b79-46ab-813b-c96c5694bc35)

**Reply by GiudGiud** [reply]
If length scale is enough to describe the interfaces then your approach works?

You can change all the properties using such if statement, one by one. 

Length scale could also be used with object I mentioned to create a new subdomain, and use new Materials that are block restricted to this new subdomain

**Reply by MinMorsh** [reply]
> If length scale is enough to describe the interfaces then your approach works?

I don't think that will help as all the interfaces have similar range of bnds value (0-0.99). 

> You can change all the properties using such if statement, one by one.

I tried that before. It doesn't help to access interfaces at different locations. I tried something like this: 

```
[Materials]
  [./ecCond] 
      type = ParsedMaterial
      block = 0
      constant_names = 'length_scale k_b k_int1 k_int2'
      constant_expressions = '1 2.84e5 3e6 5e6' # electrical conductivity of Si at 300K (S/m)
      expression = 'sk_b:= length_scale*k_b; sk_int1:= length_scale*k_int1; sk_int2:= length_scale*k_int2; if(bnds>0.9,sk_b,if(0.001<bnds<0.3,sk_int2,sk_int1))'
      outputs = exodus
      property_name = ecC
      coupled_variables = 'bnds'
    [../]
[]
```

Is there any other way to access the interfaces selectively? I can share my input file if that helps.

**Reply by GiudGiud** [reply]
how do you plan to distinguish the interfaces between each other then?
Do you mean to split them by grain type on both sides? 
Do you want every single one to be different / accessible uniquely?

**Reply by MinMorsh** [reply]
> how do you plan to distinguish the interfaces between each other then?

Is it possible to use 'volume fraction' such that I can say '40% of the interface should be SiO2 and the rest is AlO2'? I can't change the length scale cause I need a fixed number and size of grains (in my case it is 64 and 20 nm respectively). 

> Do you mean to split them by grain type on both sides? 

If I can, then yes. 

> Do you want every single one to be different / accessible uniquely?

That would be the best case scenario.

**Reply by GiudGiud** [reply]
You can use a variable to describe a fraction. The order parameters do that. The problem is that the fraction is not enough here, since you want to distinguish interface regions depending on what grains they are close to.

I think the easiest solution here will be to create a user object (or more than one, composition is often a good idea) that:
- loops over the mesh and finds the centroids of the grains (this might already exist). This could be a new "Positions" object (Positions is a system in MOOSE)
- examines the interfaces based on the values of the order parameters, and uses the grain centroids to distinguish (the two closest ones for example) between interfaces

**Reply by MinMorsh** [reply]
Thank you so much for the suggestion. Couple of follow-up questions: 

1. How can I access the grains individually in the new userobject if my order parameter is less than the number of grains (e.g. my order parameter is 20 whereas my number of grain is 64)? 

2. If I want to distinguish the interfaces just based on the volume fraction (not based on the grains), what can be the approach?

**Reply by GiudGiud** [reply]
1) this should be done by the GrainTracker already

2) similar but you ll have to do the mapping / keeping track yourself. The grain tracker keeps track of grains that's it

**Comment by PengWei97** [comment]
Hi @MinMorsh, you could potentially achieve this by utilizing a method to differentiate the interface from the bulk based on the number of activate order parameters at each integration point [_op]. Here's a suggestion on how to implement this:

1. Use GrainTracker: The GrainTracker can be used to track the grains and interfaces. In your case, you could define an interface as a region where more than one order parameter is active, indicating the presence of two grains in contact.

2. Define Interface vs. Bulk: You can set up logic to distinguish between the bulk and the interface based on the number of active order parameters at a given integration point. If only one order parameter is active, it indicates a bulk point, and if multiple order parameters are active, it indicates an interface.

3. Assign Material Properties: Once you have identified the bulk and interface regions, you can assign different material properties based on this classification. 

Let me know if you need more details or specific examples for your setup!

Wei

**Reply by MinMorsh** [reply]
Hi @PengWei97. Thanks for the suggestion. Does this method allow to apply multiple interfacial properties? I would love to discuss the details.

**Reply by PengWei97** [reply]
I think it is possible to apply multiple interface properties, but you need to be clear about the criteria you use to define different interfaces. Meanwhile, [GBAnisotropyMisori.C](https://github.com/PengWei97/mypanda/blob/prm2-GNSThermal-AGG2023/src/materials/grain_growth/GBAnisotropyMisori.C) is a class I built based on the above ideas, you can have a look.
